### PR TITLE
implement serialization API which does not rely on protobuf-java

### PR DESCRIPTION
**Pull request by richardstartin**
### What does this PR do?

This PR provides an API which allows callers to produce protobuf which can be consumed by other DDSketch implementations without needing the protobuf-java library.

### Motivation

We need to send DDSketch to other services, but can't afford the 1.6MB protobuf dependency and the ~700 classes this adds.

### CODE REVIEW

**Comment by richardstartin** [comment]
This is currently breaking on negative values occasionally so please hold off on reviewing

File: src/main/java/com/datadoghq/sketch/ddsketch/mapping/BitwiseLinearlyInterpolatedMapping.java
Comment (richardstartin) [inline src/main/java/com/datadoghq/sketch/ddsketch/mapping/BitwiseLinearlyInterpolatedMapping.java:101]: This inherits a bug fixed in #35

**Comment by richardstartin** [comment]
I separated the protobuf API into another sourceset. The user needs to add a dependency as follows if they want to use protobuf
```
    implementation('com.datadoghq:sketches-java:$version')

    runtimeOnly('com.datadoghq:sketches-java:$version') {
        capabilities {
            requireCapability("com.datadogq:protobuf-support")
        }
    }
```
The approach to deserializing a sketch is basically unchanged - swap `DDSketch.fromProto` to `DDSketchProtoBinding.fromProto` (same package).

Serialization is now done via the static method `DDSketchProtoBinding::toProto` instead of an instance method.

File: src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java
Comment (CharlesMasson) [inline src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:339]: That's a detail, but I realize we should probably have had the `minIndex` first to allow optimizations based on the value of `minIndex` that would have allowed stopping deserializing early.

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/DDSketchProtoBinding.java
Comment (CharlesMasson) [inline src/protobuf/java/com/datadoghq/sketch/ddsketch/DDSketchProtoBinding.java:15]: We could mention here that `DDSketch.serialize()` can be used.

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/mapping/IndexMappingProtoBinding.java
Comment (CharlesMasson) [inline src/protobuf/java/com/datadoghq/sketch/ddsketch/mapping/IndexMappingProtoBinding.java:32]: Should we make it public?

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java
Comment (CharlesMasson) [inline src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java:23]: Can't we now use `forEach`?
```suggestion
        store.forEach((index, count) -> storeBuilder.putBinCounts(index, count));
```

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java
Comment (CharlesMasson) [inline src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java:30]: Same here, I think we need to change.

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java
Comment (CharlesMasson) [inline src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java:33]: No need to change now, but I think I used `long` to handle the case `store.maxIndex == Integer.MAX_VALUE`, but I wonder about the cost now. Maybe should we favor what we do elsewhere, with `index < store.maxIndex` and the special handling of the last value.

File: src/test/java/com/datadoghq/sketch/ddsketch/DDSketchTest.java
Comment (CharlesMasson) [inline src/test/java/com/datadoghq/sketch/ddsketch/DDSketchTest.java:157]: Can we also test here serializing with `DDSketch.serialize()` and deserializing with `DDSketchProtoBinding .fromProto`, and possibly, the equality of the serialized byte arrays produced with `DDSketch.serialize` and `DDSketchProtoBinding.toProto`? We can also rename the method.

Same comment for `StoreTest.testProtoRoundtrip` and `LogLikeIndexMapping.testProtoRoundTrip`.

File: src/test/java/com/datadoghq/sketch/ddsketch/SerializerTest.java
Comment (CharlesMasson) [inline src/test/java/com/datadoghq/sketch/ddsketch/SerializerTest.java:74]: Nit.:
```suggestion
    private void assertEquals(Store expected, Store actual) {
```

File: src/test/java/com/datadoghq/sketch/ddsketch/SerializerTest.java
Comment (CharlesMasson) [inline src/test/java/com/datadoghq/sketch/ddsketch/SerializerTest.java:83]: Shouldn't it be `&&`? It is supposed to iterate over non-empty buckets. Alternatively, we can check that the remaining ones have counts close to zero.
```suggestion
        assertFalse(expectedIt.hasNext() && actualIt.hasNext());
```

File: src/test/java/com/datadoghq/sketch/ddsketch/SerializerTest.java
Comment (CharlesMasson) [inline src/test/java/com/datadoghq/sketch/ddsketch/SerializerTest.java:61]: We use `AccuracyTester.FLOATING_POINT_ACCEPTABLE_ERROR` in `DDSketchTest` and `StoreTest`.

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java
Comment (richardstartin) [reply src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java:33]: Good catch, I just copied and pasted this. Using `long` indexes in loops disables autovectorisation in hotspot.

File: src/main/java/com/datadoghq/sketch/ddsketch/DDSketch.java
Comment (richardstartin) [reply src/main/java/com/datadoghq/sketch/ddsketch/DDSketch.java:351]: ok

File: src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java
Comment (richardstartin) [reply src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:328]: yes I think so

File: src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java
Comment (richardstartin) [reply src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:339]: ok makes sense

Comment (richardstartin) [reply src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:339]: Although, this is pretty dangerous to rely on

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/mapping/IndexMappingProtoBinding.java
Comment (richardstartin) [reply src/protobuf/java/com/datadoghq/sketch/ddsketch/mapping/IndexMappingProtoBinding.java:32]: This is public, but in non canonical order

File: src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java
Comment (CharlesMasson) [reply src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:339]: Yes. I think it's not worth changing at this point, especially since other libs rely on the current protobuf message format.

Comment (richardstartin) [reply src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:339]: We can put it first anyway to help partial deserialisers. The keys don't need to be ordered in the message, they just need to correspond to the field numbers in the schema. 

Incidentally, for non-repeated fields, the parser is obliged to take the _last_ value matching the key; the wire format permits repeating non-repeated fields. So arguably every partial protobuf deserialisation technique I've seen is wrong, no matter how well it can work if the circle of trust is small enough.

Comment (CharlesMasson) [reply src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:339]: I see, up to you then, but I think that makes partial deserialization optimizations less interesting then, and the non-dependency on the order in the message means that we can leave that for later if the need for it really arises.

**Comment by richardstartin** [comment]
@CharlesMasson my latest commit addresses all your comments, but I would like to rebase the commits into a cohesive unit before merging

File: src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java
Comment (richardstartin) [reply src/protobuf/java/com/datadoghq/sketch/ddsketch/store/StoreProtoBinding.java]: I changed this - I think it's fine to iterate `[store.minIndex - store.offset, store.maxIndex - store.offset]` because even if the upper bound is `Integer.MAX_VALUE`, that's 4 or 5 elements larger than any JVM will let you create.

File: src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java
Comment (richardstartin) [reply src/main/java/com/datadoghq/sketch/ddsketch/store/DenseStore.java:339]: this would make testing painful though because it breaks comparisons of the binary output, let's address it another time.

**Review by CharlesMasson — APPROVED** [review]
As discussed, it'd be great to have a second pair of eyes look at the gradle changes.

Apart from that, LGTM. Thanks!

**Review by randomanderson — APPROVED** [review]
Reviewing just the Gradle changes, it looks correct to me.  The generated `pom.xml` is:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.datadoghq</groupId>
  <artifactId>sketches-java</artifactId>
  <version>0.5.1-SNAPSHOT</version>
  <name>DDSketch</name>
  <description>A Fast and Fully-Mergeable Quantile Sketch with Relative-Error Guarantees.</description>
  <url>https://github.com/DataDog/sketches-java</url>
  <licenses>
    <license>
      <name>The Apache License, Version 2.0</name>
      <url>http://www.apache.org/licenses/LICENSE-2.0</url>
    </license>
  </licenses>
  <developers>
    <developer>
      <name>Charles Masson</name>
      <email>charles.masson@datadoghq.com</email>
      <organization>Datadog</organization>
    </developer>
    <developer>
      <name>Jee E. Rim</name>
      <email>jee.rim@datadoghq.com</email>
      <organization>Datadog</organization>
    </developer>
    <developer>
      <name>Homin K. Lee</name>
      <email>homin@datadoghq.com</email>
      <organization>Datadog</organization>
    </developer>
  </developers>
  <scm>
    <connection>scm:git:git://github.com/DataDog/sketches-java.git</connection>
    <developerConnection>scm:git:ssh://github.com:DataDog/sketches-java.git</developerConnection>
    <url>http://github.com/DataDog/sketches-java</url>
  </scm>
  <dependencies>
    <dependency>
      <groupId>com.google.protobuf</groupId>
      <artifactId>protobuf-java</artifactId>
      <version>3.13.0</version>
      <scope>compile</scope>
      <optional>true</optional>
    </dependency>
  </dependencies>
</project>

```

**Comment by richardstartin** [comment]
Just squashed my commits and force pushed

**Pull request merged**
